### ISSUE TITLE
Integrate plugin with the GitHub API

### ISSUE THREAD

**Issue by RadValentin** [uberVU/github-changelog#5]
Main issue: #2 
## Need

We need to use the GitHub data people are already using to offer a transparent status update system for their customers, without building an additional admin panel for those status updates.
## Deliverables
- A method that returns titles and labels from issues that have a certain `label` and were referenced in merge commits from a given `branch`, newer than a given `timestamp`.

Ex: For `timestamp: '1400251736447'`, `branch: 'master'` and `label: 'release'` the method would go through the following steps:
- We get all commits from `master` newer than `1400251736447`, from a given repo
- We filter out commits that don't have the `release` label
- We filter out non-merge commits (maybe we have a way to figure that out, worst case by matching commit message)
- All commits have an issue reference inside their message, e.g. `Solve yata yata #11340` that basically references to `github.com/org/repo/issues/11340`
- We extract the ID from the that string, for each merge commit we're left with after applying the previous filters
- We query GitHub for each of the issue IDs we care about and return the title and all the other _labels_ it has (besides `release`, of course)

The output would then look something like this:

``` javascript
[{title: "Fix overlapping text", labels: ['bug', 'client', 'moderate']},
 {title: "Create geo map visualization", labels: ['feature']}]
```
## Solution

Given an `timestamp`, `branch` and `label` we need to build the following functionality:
1. A method to get all merge commits since the given `timestamp` (filter by `branch`) and filter out commits that don't have the `label`
2. A method to extract issue meta data from commit issue ID reference (ex: #3351)
3. A method to filter issues by `label`
4. A method to return `title` and `label` that the plugin can use to populate the notification list
#### Prerequisites
- [GitHub API](https://developer.github.com/v3/)
  - [List commits on a repository](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository)
  - [Get a single issue](https://developer.github.com/v3/issues/#get-a-single-issue)
  - [Traversing with pagination](https://developer.github.com/guides/traversing-with-pagination/)
  - [API rate limiting](https://developer.github.com/v3/#rate-limiting)
- [jQuery Plugins](http://learn.jquery.com/plugins/)
- How to access private repos through the GitHub API?
- How to insert needed credentials to access private repos?
- How to check all merged pull-requests into a branch? 
- How to filter pull-requests by timestamp?
- How to extract issue id from a pull-request?
- How to filter issues by label?
- How to get issue info for use in the plugin?
#### TODO
- [ ] Research prerequisites, answer questions
- [ ] Implement GitHub authentication
- [ ] Implement a way to check merged pull-requests
- [ ] Implement filtering by timestamp on pull-requests
- [ ] Implement a way to extract issue id from a pull-request
- [ ] Implement a way to filter issues by label

cc: @skidding @catalinmiron

**Comment by ovidiuch** [comment]
> ## Need
> 
> We need a way to get/filter all merged pull-requests on a given branch which are newer than a provided timestamp(since) and extract title and label from their corresponding issue

This is more of a Deliverable than a Need. The Need would be more like, we need to use the github data people are already using (this is the assumption this project is making. we are at least) to offer a transparent status update system for their customers, without building an additional admin panel for those status updates.

> ## Deliverable
> 
> Methods that integrate with the GitHub API and that will filter and extract titles and labels according to the provided timestamp, branch and label

Yes, but I'm not sure @mihneadb would easily figure out what we're trying to do if we were to ask for his GitHub API expertise :)

So let's see how we can rephrase it clearer. We want to deliver:

A method that returns titles and labels from issues that have a certain `label` and were referenced in merge commits from a given `branch`, newer than a given `timestamp`

E.g. A method that returns titles and lables from issues that have the label `release` and were referenced in merge commits from `master`, newer than `1400251736447`

It's still a bit hard to follow, so let's try to split it into steps (w/out knowing the GitHub API), similar to your solution.
- We get all commits from `master` newer than `1400251736447`, from a given repo
- We filter out commits that don't have the `release` label
- We filter out non-merge commits (maybe we have a way to figure that out, worst case by matching commit message)
- All commits have an issue reference inside their message, e.g. `Solve yata yata #11340` that basically references to github.com/org/repo/issues/11340
- We extract the ID from the that string, for each merge commit we're left with after applying the previous filters
- We query GitHub for each of the issue IDs we care about and return the _title_ and all the other _lables_ it has (besides `Release`, of course)

So basically input is: `branch`, `label` (to filter with, we don't want to display ALL merges) and `since` timestamp. And output is:

``` js
[{title: "Fix overlapping text", lables: ['bug', 'client', 'moderate']},
 {title: "Create geo map visualization", lables: ['feature']}]
```

_Notice we removed the #XXX id from the issue titles, we don't want clients to figure out our issue counter :)_

@mihneadb can you help us figure out if the GitHub API can provide this info?

**Comment by mihneadb** [comment]
This is definitely doable, we just need to figure out a way of optimizing the amount of API calls we do. It's worth mentioning that the approach you outlined relies on the fact that commit messages include a reference to the issue number. That's not necessarily the case across all of github. IMO it's worth thinking about going through closed issues rather than commits (the only shortcoming is that we couldn't filter by branch this way, but by milestone).

The pieces we need for what you asked:
1. there's no way of getting just merge commits, so we need to get all commits of the given branch, since the given timestamp. [docs](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository)
2. filter out the ones that have just one parent (merge commits will have two)
3. extract issue references from the commit messages
4. for every issue, get its [data](https://developer.github.com/v3/issues/#get-a-single-issue) and extract the title and labels

Numbers:
- you can get up to **100** commits in a "page" (response of API call) [1]
- you can make **60** _unauthenticated_ requests _per hour_ [2]
- you can make **5000** _authenticated_ requests _per hour_ [2]

[1] https://developer.github.com/guides/traversing-with-pagination/
[2] https://developer.github.com/v3/#rate-limiting

**Alternative solution**:
(If we are ok with filtering by milestone instead of branch.)
1. Get the closed issues. We can specify a timestamp, a milestone, state=closed etc. [docs](https://developer.github.com/v3/issues/#list-issues-for-a-repository)
2. For every issue, extract the title and labels.

This approach is cheaper in terms of API calls, because we will probably get away with just one call on average.

BTW, I can write the code for this if it's fine for you, since I've already interacted with the API quite a bit. :)

**Comment by ovidiuch** [comment]
This sounds great. We already plan to filter the visible issues by a "release" label (if the api doesn't allow we can do it on our side), so sure, we can query issues directly. Our deploy system is hooked to github merges, so this is the order of events
- Pull request is merged
- Issue is closed (either 1. because pull was issue itself, opened with hub, or 2. because pull had "fixed #ISSUE_ID" body
- We receive github event and start deploy
- Deploy is finished and interface receives "update" message
- ... we can now query issues, as they were clearly closed by now (the only corner-case problem would be with concurring deploys but we don't support that)

Another thing to consider is the visible API KEY (thanks to @claudiu-coman for raising this problem). While I would like the jQuery plugin to support this out of the box (and luckily the GitHub api supports JSON-P), we should clearly warn users that it should be hidden on the server side, and maybe provide out-of-the-box .py/.php/etc proxies that receive the same payloads as GitHub methods but without API KEY, then append API KEY stored inside those proxy files as a simple config, and forward the requests to GitHub and response back. wdyt?

Sure, it would be awesome if you contributed :)

**Comment by aismail** [comment]
I think we don't need to build a proxy ourselves. My intuition tells me there should be some services who expose that to you already (not sure how you authenticate to them, though).

**Comment by mihneadb** [comment]
Indeed, there is the problem of the API key. Without a proxy, someone could figure out the API key and get access to info on `thehole`, for example.

We can look for services like @aismail mentioned. I checked a bit and came up empty. I don't think the user that wants to view the notifications has to be authenticated.

Obviously, building such proxies ourselves wouldn't be that hard, they only have to add the key to the headers, forward the request and then reply.

**Comment by ovidiuch** [comment]
> Indeed, there is the problem of the API key. Without a proxy, someone could figure out the API key and get access to info on thehole, for example.
> 
> We can look for services like @aismail mentioned. I checked a bit and came up empty. I don't think the user that wants to view the notifications has to be authenticated.

Don't see why add any layer of complexity through authentication / dependency on 3rd party services. This is meant to work out-of-the-box for public projects.

> Obviously, building such proxies ourselves wouldn't be that hard, they only have to add the key to the headers, forward the request and then reply.

Exactly.

@mihneadb do you see any problem with my previous comment?

**Comment by mihneadb** [comment]
On May 20, 2014 11:44 AM, "Ovidiu Cherecheș" notifications@github.com
wrote:

> > Indeed, there is the problem of the API key. Without a proxy, someone
> > could figure out the API key and get access to info on thehole, for example.
> > 
> > We can look for services like @aismail mentioned. I checked a bit and
> > came up empty. I don't think the user that wants to view the notifications
> > has to be authenticated.
> 
> Don't see why add any layer of complexity through authentication /
> dependency on 3rd party services. This is meant to work out-of-the-box for
> public projects.

Indeed, if we release it to the public it would make things harder. We
should also allow people to use it with no authentication, this will make
it super easy for open source projects, no proxy required and 60 requests
should be enough per hour.

> > Obviously, building such proxies ourselves wouldn't be that hard, they
> > only have to add the key to the headers, forward the request and then reply.
> 
> Exactly.
> 
> @mihneadb do you see any problem with my previous comment?

No, I don't. I'll come up with the JS today in a codepen or something so we
can play with it.

## 

Mihnea Dobrescu-Balaur
Sent from a mobile device.

**Comment by mihneadb** [comment]
Here's a quick demo: http://codepen.io/mihneadb/pen/yhrHi
The API lets us filter before querying by label(s). We also get back the colors of the labels (among other things), so we can keep the UI consistent.

cc @skidding

**Comment by ovidiuch** [comment]
Something seems broken

![image](https://cloud.githubusercontent.com/assets/250750/3026647/3ae61884-e018-11e3-89d6-bef6d7bc101f.png)

**Comment by mihneadb** [comment]
That's funny, it loads fine in FF but not in Chrome. Must be something that Codepen doesn't like about it. I made a gist with the whole thing: https://gist.github.com/mihneadb/2c4ef6a9ee0e2c092ea3

**Comment by RadValentin** [comment]
Ran into the same issue on my end, it's because "arrow functions" aren't supported in Chrome yet, [MDN link](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/arrow_functions#Browser_compatibility). I'm sure there's an easy way to get around this though.

**Comment by mihneadb** [comment]
Oh, silly me. I'll fix that, just a sec.

**Comment by mihneadb** [comment]
There we go: http://codepen.io/mihneadb/pen/yhrHi?editors=001
(don't mind the looks/style, just demo-ing the functionality :) ).

**Comment by ovidiuch** [comment]
Awesome, it looks like we're on to something!

The rest are details:
- Updating the since timestamp after each request to prevent duplication
- Removing #XXXX suffix because we don't want users to see our issue count (If only we use this and is not common practice we can use a fork for us internally)
- etc.

**Comment by mihneadb** [comment]
Yes, those are all doable. BTW, I'm thinking of using `localStorage` for things like updating the `since` timestamp and maybe caching. It's well [supported](http://caniuse.com/#search=localstorage). Any objections?

**Comment by ovidiuch** [comment]
Why need to persist anything? The concept behind these updates is that they vanish when you refresh. Notifications are shown for updates that occurred between your last hard refresh and the last time the plugin queried GitHub.

**Comment by mihneadb** [comment]
Ah, ok. I thought we'd like to show the old "read" ones as well, kind of like Facebook's notification box.
I guess I must be missing something but how will we keep the value of `since` between hard refreshes? Let's say that user was on the page today. They close the page and tomorrow they load a fresh version. How will we know what `since` to use (without some sort of persistence layer)?

**Comment by ovidiuch** [comment]
We don't show _read_ ones in the first iteration. `since` basically starts as `Date.now()` and continues update as `Date.now()` as we do refreshes. If you close the page the updates list will always be clean (missing altogether with our design) when you open it again.

**Comment by mihneadb** [comment]
Ok. I'll wait for #8 to get merged and then I'll submit a PR to replace the dummy data, and we can iterate on that.

**Comment by ovidiuch** [comment]
We managed to write a pretty robust jQuery plugin that already works out of the box. Can you take a look at https://github.com/uberVU/github-changelog/pull/8 at provide some feedback?

PS. You can test the plugin in action by pulling that branch `2--changelog-jquery-plugin-main`, opening index.html in your browser and opening then closing an issue on github that has the labels "demo" and on of the three "bug", "feature" or "enhancement".

Thanks!

**Comment by mihneadb** [comment]
Looks good, tested and it works. :)

Only question would be why are we clearing and re-setting the `checkForUpdates` interval? Is it to support changing refresh intervals on the fly (i.e. without reinstantiating)?

**Comment by ovidiuch** [comment]
A different way to rephrase the question would be why setTimeout and not setInterval, right? I guess the main motivation (5-sec judgement call :p) was that it's easier to reset the timer when you call `checkForUpdates` by hand. In such a way that the next interval is X seconds after the manual call, and not the previous interval. Hope that makes sense.

**Comment by mihneadb** [comment]
Oh I see where my confusion comes from: we use setTimeout and clear it with
clearInterval. :)

**Comment by ovidiuch** [comment]
Epic. Will correct that, weird that it works :))

**Comment by mihneadb** [comment]
Because browsers! :))

## 

Mihnea Dobrescu-Balaur

**Comment by ovidiuch** [comment]
Integration MVP is working, closing this for now.
